### PR TITLE
RNN fusion across layers 

### PR DESCRIPTION
**Pull request by pruthviIntel**
- Added graph pass for fusing RNN op across layer
- Added test case for inter v/s cpu for verifying layer fused RNN
- more sanity checks in the RNN fusion graph pass
- added support to replace the recurrent cell state correctly in the fused RNN op

### CODE REVIEW

**Comment by pruthviIntel** [comment]
@jbobba @Krovatkin @louisfeng ping for review :)

File: src/ngraph/runtime/cpu/pass/cpu_rnn_fusion.cpp [hunk 1]
```diff
@@ -242,10 +242,12 @@ void ngraph::runtime::cpu::pass::LSTMFusion::construct_lstm_fprop()
         auto ht_output = std::make_shared<op::GetOutputElement>(lstm, 0);
         auto ct_output = std::make_shared<op::GetOutputElement>(lstm, 1);
 
+        if (lstm->get_outputs().at(0).get_inputs().size() != 2)
+        {
+            throw ngraph_error("Lstm node doesnt have two outputs");
```
Comment (louisfeng) [inline src/ngraph/runtime/cpu/pass/cpu_rnn_fusion.cpp]: The error message was a bit ambiguous to me at first look, maybe something like:
`throw ngraph_error("Lstm node requires two outputs");`?

File: test/cpu_fusion.cpp [hunk 2]
```diff
@@ -2076,3 +2077,43 @@ TEST(cpu_fusion, sigmoid_multiply_fusion_backward)
...
+    const string json_string = file_util::read_file_to_string(json_path);
+    stringstream ss(json_string);
+    shared_ptr<Function> func = ngraph::deserialize(ss);
+    pass_manager.run_passes(func);
+    size_t ref_rnn_count = 1;
+    auto rnn_count = count_ops_of_type<op::Rnn>(func);
+    EXPECT_EQ(ref_rnn_count, rnn_count);
+}
+
+TEST(cpu_fusion, fuse_rnn_across_2layer_1timestep)
```
Comment (louisfeng) [inline test/cpu_fusion.cpp]: Do you have a test with multiple layer and multiple time steps?

Comment (pruthviIntel) [reply test/cpu_fusion.cpp]: i dont have the serialized graph where the feature size is same for Xt and ht for multiple time steps (which is the requirement for fusing across layer).  I hacked mxnet unit test to get the  current serialized graph. but for some reason i am not able to unroll those cells in the serialized graph. So each RNN layer have just 1 cell. I will try to get that.

**Comment by pruthviIntel** [comment]
@louisfeng can u complete the review? :). @jbobba @Krovatkin one more ping for review :)

File: src/ngraph/runtime/cpu/pass/cpu_rnn_fusion.hpp [hunk 3]
```diff
@@ -29,6 +29,7 @@ namespace ngraph
             {
                 class LSTMFusion;
                 class RNNFusion;
+                class RecurrentRNNFusion;
```
Comment (jbobba) [inline src/ngraph/runtime/cpu/pass/cpu_rnn_fusion.hpp]: Rename to MultiLayerRNNFusion.

File: test/cpu_fusion.cpp [hunk 4]
```diff
@@ -2075,3 +2076,45 @@ TEST(cpu_fusion, fuse_batch_dot_forward)
         EXPECT_TRUE(test::all_close(cpu_results.at(i), int_results.at(i), 1.0e-4f, 1.0e-4f));
     }
 }
+
+TEST(cpu_fusion, fuse_rnn_across_layer)
+{
+    pass::Manager pass_manager;
+    pass_manager.register_pass<runtime::cpu::pass::LSTMFusion>();
+    pass_manager.register_pass<runtime::cpu::pass::RNNFusion>();
+    pass_manager.register_pass<ngraph::pass::AlgebraicSimplification>();
```
Comment (jbobba) [inline test/cpu_fusion.cpp]: Why do we need algebraic simplification pass here? Add comment.

Comment (pruthviIntel) [reply test/cpu_fusion.cpp]: when we fuse RNN and feed its output to the next RNN layer. we have the below slice->concat thing.
screen shot 2018-06-12 at 10 39 58 am

One of the registered pass in Algebraic simplification pass basically looks in to {GOE->slices->concat-> RNN} and replace them with {GOE->RNN}, this helps to simply the RNN multi layer fusion pass.

File: src/ngraph/runtime/cpu/pass/cpu_rnn_fusion.hpp (hunk 3 above)
Comment (pruthviIntel) [reply src/ngraph/runtime/cpu/pass/cpu_rnn_fusion.hpp]: will do

File: src/ngraph/runtime/cpu/pass/cpu_rnn_fusion.cpp [hunk 5]
```diff
@@ -588,3 +625,213 @@ void ngraph::runtime::cpu::pass::RNNFusion::construct_rnn_lstm_fprop()
...
+        // the intermediate inputs for the next layer will be computed by the MKLDNN
+        auto src_layer_nodes = m.get_bound_nodes_for_pattern(src_layer_label);
+        auto src_layer = src_layer_nodes[src_layer_nodes.size() - 1];
+
+        auto src_iter = compute_multi_layer_rnn_inputs(src_iter_label, m);
+        auto weights_layer = compute_multi_layer_rnn_inputs(weights_layer_label, m);
+        auto weights_iter = compute_multi_layer_rnn_inputs(weights_iter_label, m);
+        auto bias = compute_multi_layer_rnn_inputs(bias_label, m);
+
+        std::shared_ptr<op::Rnn> rnn_node = nullptr;
```
Comment (Krovatkin) [inline src/ngraph/runtime/cpu/pass/cpu_rnn_fusion.cpp]: this is the last rnn node (i.e. closest to `op::Result`?)? maybe we could add a comment here?

File: src/ngraph/runtime/cpu/pass/cpu_rnn_fusion.cpp [hunk 6]
```diff
@@ -588,3 +625,213 @@ void ngraph::runtime::cpu::pass::RNNFusion::construct_rnn_lstm_fprop()
...
+                                             num_gates_in_lstm,
+                                             sequence_len,
+                                             src_layer_feature_size,
+                                             feature_size,
+                                             num_rnn_cell_states,
+                                             rnn_direction,
+                                             num_fused_rnn_layers);
+
+        auto layer_rnn_ht = std::make_shared<op::GetOutputElement>(rnn, 0);
+        auto layer_rnn_ht_ct = std::make_shared<op::GetOutputElement>(rnn, 1);
```
Comment (Krovatkin) [inline src/ngraph/runtime/cpu/pass/cpu_rnn_fusion.cpp]: is it really `ht_ct`?

Comment (pruthviIntel) [reply src/ngraph/runtime/cpu/pass/cpu_rnn_fusion.cpp]: yes, the second output of MKLDNN RNN is dense output {ht|ct}

GOE0 -> {ht}
GOE1 -> {ht | ct}

**Pull request merged**